File: java/tema_4/src/main/java/ro/mpp2024/concertpro/dao/repository/user_repository/UserDbRepository.java
package ro.mpp2024.concertpro.dao.repository.user_repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.concertpro.dao.model.User;
import ro.mpp2024.concertpro.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Properties;

public class UserDbRepository implements IUserRepository {
    private final JdbcUtils jdbcUtils;
    private static final Logger logger = LogManager.getLogger();

    public UserDbRepository(Properties props) {
        logger.info("Initializing UserDbRepository with properties: {} ", props);
        this.jdbcUtils = new JdbcUtils(props);
    }

    @Override
    public User findOne(Long userId) {
        return null;
    }

    @Override
    public Iterable<User> findAll() {
        return null;
    }

    @Override
    public void save(User user) {
    }

    @Override
    public void update(User user, Long userId) {
        logger.traceEntry("update task {} ", user);
        Connection connection = this.jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("update users set username = ?, " +
                "password = ?, loggedIn = ? where userId = ?")) {
            preparedStatement.setString(1, user.getUsername());
            preparedStatement.setString(2, user.getPassword());
            preparedStatement.setBoolean(3, user.getLoggedIn());
            preparedStatement.setLong(4, userId);

            int result = preparedStatement.executeUpdate();
            logger.trace("Saved {} instances", result);
        } catch (SQLException exception) {
            logger.error(exception);
            System.err.println("Error DB: " + exception);
        }

        logger.traceExit();
    }

    @Override
    public void delete(Long userId) {
    }

    @Override
    public User findByUsernameAndPassword(String username, String password) {
        logger.traceEntry();
        Connection connection = this.jdbcUtils.getConnection();
        User user = new User();

        try(PreparedStatement preparedStatement = connection.prepareStatement("SELECT * FROM users WHERE " +
                "username = ? AND password = ?")) {
            preparedStatement.setString(1, username);
            preparedStatement.setString(2, password);

            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                while(resultSet.next()) {
                    Long userId = resultSet.getLong("userId");
                    String usernameN = resultSet.getString("username");
                    String passwordN = resultSet.getString("password");
                    Boolean loggedIn = resultSet.getBoolean("loggedIn");

                    user.setId(userId);
                    user.setUsername(usernameN);
                    user.setPassword(passwordN);
                    user.setLoggedIn(loggedIn);
                }
            }
        } catch (SQLException exception) {
            logger.error(exception);
            System.err.println("Error DB: " + exception);
        }

        logger.traceExit();
        return user;
    }
}

File: c#/tema_5/Persistence/UserRepository/UserDbRepository.cs
using System.Collections.Generic;
using System.Data;
using Model;
using Persistence.DbUtils;

namespace Persistence.UserRepository
{
    public class UserDbRepository : IUserRepository
    {
        // private static readonly ILog Log = LogManager.GetLogger("SpectacleDbRepository");
        private readonly IDictionary<string, string> _properties;

        public UserDbRepository(IDictionary<string, string> properties)
        {
            _properties = properties;
        }

        public User FindOne(long id)
        {
            throw new System.NotImplementedException();
        }

        public IEnumerable<User> FindAll()
        {
            IDbConnection connection = Persistence.DbUtils.DbUtils.GetConnection(_properties);
            IList<User> users = new List<User>();

            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "select * from users";

                using (IDataReader dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        long userId = dataReader.GetInt64(0);
                        string username = dataReader.GetString(1);
                        string password = dataReader.GetString(2);
                        bool loggedIn = dataReader.GetBoolean(3);

                        User user = new User(username, password, loggedIn);
                        user.Id = userId;
                        users.Add(user);
                    }
                }
            }

            return users;
        }

        public void Save(User entity)
        {
            throw new System.NotImplementedException();
        }

        public void Update(User entity, long id)
        {
            // Log.InfoFormat("Update User Task {0}", entity);
            IDbConnection connection = DbUtils.DbUtils.GetConnection(_properties);

            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "update users set username = @username, password = @password, " +
                                      "loggedIn = @loggedIn where userId = @userId";
                
                IDbDataParameter usernameDataParameter = command.CreateParameter();
                usernameDataParameter.ParameterName = "@username";
                usernameDataParameter.Value = entity.Username;
                command.Parameters.Add(usernameDataParameter);
                
                IDbDataParameter passwordParameter = command.CreateParameter();
                passwordParameter.ParameterName = "@password";
                passwordParameter.Value = entity.Password;
                command.Parameters.Add(passwordParameter);
                
                IDbDataParameter loggedInParameter = command.CreateParameter();
                loggedInParameter.ParameterName = "@loggedIn";
                loggedInParameter.Value = entity.LoggedIn;
                command.Parameters.Add(loggedInParameter);
                
                IDbDataParameter userIdParameter = command.CreateParameter();
                userIdParameter.ParameterName = "@userId";
                userIdParameter.Value = id;
                command.Parameters.Add(userIdParameter);

                int result = command.ExecuteNonQuery();
                if (result == 0)
                {
                    // Log.Error("Error DB: No User Updated");
                }
                
                // Log.InfoFormat("Saved {0} instances", result);
            }
        }

        public void Delete(long id)
        {
            throw new System.NotImplementedException();
        }

        public User FindByUsernameAndPassword(string username, string password)
        {
            // Log.InfoFormat("Entering FindUserByUsernameAndPassword with values {0}, {1}", username, password);
            IDbConnection connection = DbUtils.DbUtils.GetConnection(_properties);

            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "select * from users where username = @username and password = @password";
                IDbDataParameter usernameParameter = command.CreateParameter();
                usernameParameter.ParameterName = "@username";
                usernameParameter.Value = username;
                command.Parameters.Add(usernameParameter);
                
                IDbDataParameter passwordParameter = command.CreateParameter();
                passwordParameter.ParameterName = "@password";
                passwordParameter.Value = password;
                command.Parameters.Add(passwordParameter);

                using (IDataReader dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        long userId = dataReader.GetInt64(0);
                        string usernameN = dataReader.GetString(1);
                        string passwordN = dataReader.GetString(2);
                        bool loggedIn = dataReader.GetBoolean(3);

                        User user = new User(usernameN, passwordN, loggedIn);
                        user.Id = userId;
                        
                        // Log.InfoFormat("Existing FindUserByUsernameAndPassword with value {0}", user);
                        return user;
                    }
                }
            }
            
            // Log.InfoFormat("Existing FindUserByUsernameAndPassword with value {0}", null);
            return null;
        }
    }
}
